### PR TITLE
vulkan: some validation error fixes

### PR DESCRIPTION
**Pull request by alichraghi**
- [x] By selecting this checkbox, I agree to license my contributions to this project under the license(s) described in the LICENSE file, and I have the right to do so or have received permission to do so by an employer or client I am producing work for whom has this right.

### CODE REVIEW

File: src/shader/codegen/msl.zig [hunk 1]
```diff
@@ -531,7 +534,7 @@ fn emitStatement(msl: *Msl, inst_idx: InstIndex) error{OutOfMemory}!void {
         .discard => try msl.emitDiscard(),
         // .@"break" => try msl.emitBreak(),
         .@"continue" => try msl.writeAll("continue;\n"),
-        // .call => |inst| try msl.emitCall(inst),
+        .call => |inst| try msl.emitCall(inst),
```
Comment (pdoane) [inline src/shader/codegen/msl.zig:537]: I think this should be removed, accidental dead-code.  Will be handled by emitExpr below?

**Review by pdoane — APPROVED** [review]

File: src/shader/codegen/msl.zig (hunk 1 above)
Comment (alichraghi) [reply src/shader/codegen/msl.zig:537]: No. function calls can be appeared as both statements and expressions.

Comment (pdoane) [reply src/shader/codegen/msl.zig:538]: Oops - sorry, I forgot the emitExpr case is not an `else`... emitCall doesn't include a `;` though so not sure how this works.  I think it should just be another case prong for line 541.

File: src/d3d12.zig [hunk 2]
```diff
@@ -471,10 +471,29 @@ pub const Device = struct {
...
+        _ = device;
+        const module = try allocator.create(ShaderModule);
+        module.* = .{ .code = .{ .code = code } };
+        return module;
+    }
+
+    pub fn createShaderModuleMSL(
+        device: *Device,
+        code: []const u8,
+        workgroup_size: sysgpu.ShaderModule.WorkgroupSize,
```
Comment (pdoane) [inline src/d3d12.zig:491]: Maybe should be optional as it's only needed for compute?

Comment (alichraghi) [reply src/d3d12.zig:491]: good catch!

File: src/shader/codegen/msl.zig (hunk 1 above)
Comment (alichraghi) [reply src/shader/codegen/msl.zig:538]: Hmm, afaik function calls like `myFn();` are statements so `;` must be added in `emitStatement`

File: src/d3d12.zig (hunk 2 above)
Comment (alichraghi) [reply src/d3d12.zig:491]: but anyway it's a little awkward, i think it's time to start refactoring the API

**Comment by pdoane** [comment]
I think my comment got lost off the PR - the new code for `call` is fine, but you could simplify it with the other expression statements and just use `.call,` and let it process as an expression with a semicolon.

**Comment by alichraghi** [comment]
Sorry i think i don't understand this. Can you show the code diff?

**Comment by pdoane** [comment]
I'm recommending this:

```
        .call,
        .assign,
        .nil_intrinsic,
        .texture_store,
        => {
            try msl.emitExpr(inst_idx);
            try msl.writeAll(";\n");
        },
```

**Comment by alichraghi** [comment]
O_O I went too deep

**Comment by slimsag** [comment]
merging to unblock other work, will test on macOS this weekend.

**Pull request merged**